### PR TITLE
Automatically build TOCs for pages

### LINKED ISSUE #894: Integrate an automated TOC generator
Find and implement optimal Jekyll extension for auto-generating the sidebar TOC from rendered page content (or source). Nothing I've seen so far is working out of the box, and this is a notorious sticking point for Markown-based Jekyll sites. 

Currently, the Table of Contents (ToC) on extension pages (formerly known as Plugin pages) are partly auto-built, and partly built "by hand".

1. For example, have a look at the right-hand nav on https://docs.konghq.com/plugins/key-authentication/
2. Then, look at the code that specifies the lower part of that ToC https://raw.githubusercontent.com/Kong/docs.konghq.com/master/app/plugins/key-authentication.md
3. Note that there is a brittle association between the nav: code at the top of the document, and the names of the headers lower in the document - if there isn't a perfect match, the ToC won't correctly link to the matching section in the document.

In the future, the full ToC should be auto-built from the headers in the document. In this arrangement, headers in a doc like this:

```
## Hello
### More here
### And yet more
```
Would render a ToC like this

**[Hello](url)**
[More here](url)
[And yet more here](url)

### PR DESCRIPTION
**Pull request by allejo**
### Summary

Greetings @coopr and Kong Doc maintainers!

Create an issue and you shall receive, especially during Hacktoberfest 🎃 I'm the guy lacking sanity who built [jekyll-toc](https://github.com/allejo/jekyll-toc), a GitHub Pages compatible way of building table of contents automagically ✨

Before continuing any further work, to save both our times I just wanted to present a quick proof of concept of how this would work.

`jekyll-toc` works by reading in HTML and extracting out H1-H6 headings, getting their IDs, and then building a TOC in markdown with that information and then `markdownify`ing it.

Since it seems that "Terminology" and "Configuration" are automatically rendered inside of the `extensions.html` layout, we'll need to combine that with `{{ content }}` and then feed that into `toc.html` so that it may generate all the headings correctly; this is being achieved by using a `{% capture %}`.

If this is something that interests you, I'd be more than happy to continue working on this but I would need *a lot* of help checking pages to make sure things are being rendered correctly. **And** checking that headings are in correct order. (e.g. ["Usage" is an H3 instead of H2 on the ACL page](https://docs.konghq.com//hub/kong-inc/acl/))

### Full changelog

* Added cool new automated ToC

### Issues resolved

Fix #894 

### Checklist:

- [ ] [Commit message & atomicity](https://github.com/Kong/docs.konghq.com/blob/master/CONTRIBUTING.md#commit-atomicity) checked
- [x] Documentation 
- [x] Spellchecked my updates
- [ ] Ready to be merged

### CODE REVIEW

**Comment by coopr** [comment]
Wow, this is super-awesome of you @allejo - and yay for Hacktoberfest!!

I'm 110% on-board with doing all the work of checking each and every page, adjusting heading levels, etc. 

One question/request: #894 perhaps implies incorrectly that this solution should apply only to pages within /hub - what would you think about making this solution apply to any page in the docs.KongHQ.com site? 

For example, you can see a Table of Contents on https://docs.konghq.com/0.14.x/secure-admin-api/ - but when you look at https://raw.githubusercontent.com/Kong/docs.konghq.com/master/app/0.14.x/secure-admin-api.md you see that the ToC is built and maintained by hand - if anyone edits/adds/removes a header in the content, they need to know and remember to also edit it in the ToC section.

**Comment by allejo** [comment]
> One question/request: #894 perhaps implies incorrectly that this solution should apply only to pages within /hub - what would you think about making this solution apply to any page in the docs.KongHQ.com site?

Sure! `jekyll-toc` can take whatever syntactically sane-ish HTML you toss at it. Since all that content is coming from markdown, it'll be sane, so we would just need to include `toc.html` in all of the layouts you'd like. I just did the `/hub` layout first since that's what was mentioned but this can easily be implemented anywhere else.

In your example, since the TOC is in the middle of the content, we could add a directive like `<!-- toc -->` in the markdown files and in the respective layout, Jekyll can look for that directive and replace it with whatever `toc.html` generates.

**tl;dr**: Yes, it's doable

**Comment by coopr** [comment]
Nice progress!

> > I don't see the new ToCs on pages for Kong Enterprise extensions like http://localhost:3000/hub/kong-inc/oauth2-introspection nor on third-party extension pages like http://localhost:3000/hub/peter-evans/paseto
> 
> Ah this was because I had left in a condition to check for `page.nav`; I removed it, so they should be generated now.

Confirmed, thanks for adjusting that!

> * Which layouts will need a toc?

I believe it is just `app/_layouts/extension.html` and `app/_layouts/docs.html`

Some of the remaining layouts are "dead code" that'll be removed soon, and I believe the other layouts like `default.html` and `page.html` are too general to need ToCs. We certainly don't need ToCs on pages like https://docs.konghq.com/ and https://docs.konghq.com/enterprise/

> * For pages under the 0.X.Y folders, is there a specific pattern as to where those tocs will be placed? For example, always before the second heading?

I see how we've been inconsistent where we've placed the manually-created ToCs. Let's keep this simple - ToCs should go at the top of the document for docs pages like https://docs.konghq.com/enterprise/0.33-x/getting-started/adding-your-api/ and https://docs.konghq.com/0.14.x/network/, and at the top of the right-hand sidebar for Hub pages like https://docs.konghq.com/hub/kong-inc/request-termination/

> * CSS changes to the heading sizes will need to be made. Because it seems like `h4` tags are being used as they generate smaller font sizes than `h3`. I designed `toc.html` to enforce proper ordering of headings so, this may cause problems if headings are out of order.

Sounds good, I'll make sure those CSS changes get done!

**Comment by coopr** [comment]
Wow, you went above and beyond with https://github.com/Kong/docs.konghq.com/pull/920/commits/8539d5d468dc5b36ddf22c71f6a738560a4ef135 - thanks @allejo ! I'm reviewing this PR now...

**Comment by coopr** [comment]
New issues:

Now these pages have auto-generated ToCs, and they should not:
http://localhost:3000/
http://localhost:3000/enterprise/

Some pages now have mal-formed ToCs, like:
http://localhost:3000/0.14.x/loadbalancing/

which looks like:

loadbalancing_reference_-_v0_14_x___kong_-_open-source_api_management_and_microservice_management

**Comment by allejo** [comment]
> New issues:
> 
> Now these pages have auto-generated ToCs, and they should not:
> [http://localhost:3000](http://localhost:3000/)
> [http://localhost:3000/enterprise](http://localhost:3000/enterprise/)

Oops... That was not my intention, let me look into that.

> Some pages now have mal-formed ToCs, like:
> [http://localhost:3000/0.14.x/loadbalancing](http://localhost:3000/0.14.x/loadbalancing/)

So this is what I was mentioning about having headings out of order, the headings will need to be fixed to be in proper order (which is what I recommend for accessibility purposes). however, we could work around this and make the TOC generator accept incorrect orders.

**Comment by coopr** [comment]
> > Some pages now have mal-formed ToCs, like:
> > [http://localhost:3000/0.14.x/loadbalancing](http://localhost:3000/0.14.x/loadbalancing/)
> 
> So this is what I was mentioning about having headings out of order, the headings will need to be fixed to be in proper order (which is what I recommend for accessibility purposes). however, we could work around this and make the TOC generator accept incorrect orders.

Ohhhh, now I get it - thanks for helping me along :) 

Ok, let me see if I can fix the heading order on that specific page, to see if it fixes the malformed ToC...

**Comment by coopr** [comment]
Ah, got it! Now I see how the heading levels have to be in order for the ToC to work - so this will cause issues:

```
# Heading
## Sub-heading
#### Sub-sub-sub-heading
```

But this will work well:

```
# Heading
## Sub-heading
### Sub-sub-heading
```

I've tuned-up the heading levels on http://localhost:3000/0.14.x/loadbalancing and now it works well, except for one thing - the bolding of the entries is inconsistent:

loadbalancing_reference_-_v0_14_x___kong_-_open-source_api_management_and_microservice_management

I would imagine we'd want H1 to be bold, and the others not bold, or maybe have all heading levels not be bold in the ToC?

**Comment by coopr** [comment]
Uh oh, I've found yet another issue - sorry for the rabbit hole here @allejo - have a look at http://localhost:3000/enterprise/changelog/ - you'll see that because we have multiple headers in that page with the same name (eg. there are multiple `### Fixes`) the ToC doesn't link correctly to the relevant heading. Any ideas?

**Comment by allejo** [comment]
I just pushed a change to sanitize the headings automatically. However, this is up to you on how you'd want to handle this. In the case of the load balancing document, your headings look like:

```
# **Blue-Green Deployments**
```

So the bolding was just being kept, by sanitizing it any custom HTML is being stripped out. The only situation where this wouldn't be ideal is if you have headings with `<code>` blocks like on http://localhost:3000/0.14.x/proxy/

---

> Uh oh, I've found yet another issue - sorry for the rabbit hole here @allejo - have a look at http://localhost:3000/enterprise/changelog/ - you'll see that because we have multiple headers in that page with the same name (eg. there are multiple ### Fixes) the ToC doesn't link correctly to the relevant heading. Any ideas?

Oh boy... this is would be a kramdown issue as the toc generator doesn't do anything about generating IDs; it just uses them. Let me investigate a bit further on this one.

Worst case, you could [explicitly specify the IDs for the headings](https://kramdown.gettalong.org/syntax.html#specifying-a-header-id), e.g.

```
# 0.33.0

## Fixes {#0-33-0-fixes}
```

**Comment by coopr** [comment]
> I just pushed a change to sanitize the headings automatically. However, this is up to you on how you'd want to handle this. In the case of the load balancing document, your headings look like:
> 
> ```
> # **Blue-Green Deployments**
> ```
> So the bolding was just being kept, by sanitizing it any custom HTML is being stripped out. The only situation where this wouldn't be ideal is if you have headings with `<code>` blocks like on http://localhost:3000/0.14.x/proxy/

Ah, I see - ya know, past us was doing silly things with headers that include bolding. I'll commit to removing all the silly formatting like the bolding, and let's remove the sanitizing so we can keep intentional things like the `<code>` blocks in headings. 


> > Uh oh, I've found yet another issue - sorry for the rabbit hole here @allejo - have a look at http://localhost:3000/enterprise/changelog/ - you'll see that because we have multiple headers in that page with the same name (eg. there are multiple ### Fixes) the ToC doesn't link correctly to the relevant heading. Any ideas?
> 
> Oh boy... this is would be a kramdown issue as the toc generator doesn't do anything about generating IDs; it just uses them. Let me investigate a bit further on this one.
> 
> Worst case, you could [explicitly specify the IDs for the headings](https://kramdown.gettalong.org/syntax.html#specifying-a-header-id), e.g.
> 
> ```
> # 0.33.0
> 
> ## Fixes {#0-33-0-fixes}
> ```

That doesn't seem like so bad of a worst case, but it'd sure be nicer if ToC generating took care of this automagically - I appreciate your investigating :)

**Comment by coopr** [comment]
> Ah scratch that thought, you're using redcarpet instead of kramdown for your markdown engine. Duplicate IDs are a known issue in redcarpet, see [vmg/redcarpet#307 (comment)](https://github.com/vmg/redcarpet/issues/307#issuecomment-261793668). The syntax in my previous post was relating to kramdown, don't know if it'll work in redcarpet. kramdown handles duplicate IDs correctly, fwiw
> 
> Jekyll moved to kramdown as the official markdown engine in Jekyll 3.0, not sure if you'd like to follow suit. You'd have to look through the docs to make sure no special markup unique to redcarpet breaks if you switch to kramdown.

I'd love for us to be on a modern version of Jekyll, along with using kramdown and surely other things we've neglected for a while - but I'm guessing that doing that upgrade as part of this particular PR is not the right approach.

**Comment by coopr** [comment]
At first I thought the "TOC isn't loaded on index.md files" would be a problem, because we _do_ want the ToC on Hub pages like https://github.com/Kong/docs.konghq.com/tree/master/app/_hub/peter-evans/paseto/index.md - but somehow even though that page is `index.md`, I see a generated ToC at http://localhost:3000/hub/peter-evans/paseto/ - could you help me understand how that happens? (I'm _happy_ that it happens, I just don't understand _why_ it happens :)

I tried out `toc: false` on that same page ^^ and it didn't make any difference - the auto-generated ToC is still present.

I tried it out on a different page http://localhost:3000/0.14.x/auth/ and it also had no effect - changing the front-matter of that file to:

auth_md_ ___documents_github_docs_konghq_com

Causes the page to render as:

authentication_reference_-_v0_14_x___kong_-_open-source_api_management_and_microservice_management

> On another topic, do you want me to rewrite the commit messages to follow a specific convention?

Our commit message guidelines are here https://github.com/Kong/docs.konghq.com/blob/master/CONTRIBUTING.md#commit-message-format thanks!

**Comment by allejo** [comment]
> At first I thought the "TOC isn't loaded on index.md files" would be a problem, because we do want the ToC on Hub pages like /app/_hub/peter-evans/paseto/index.md@master - but somehow even though that page is index.md, I see a generated ToC at http://localhost:3000/hub/peter-evans/paseto - could you help me understand how that happens? (I'm happy that it happens, I just don't understand why it happens :)

So, I'm checking for `index.md` [only in the `docs.html` layout](https://github.com/Kong/docs.konghq.com/pull/920/files#diff-d2c714758762d35c4930e19c88f583b3R122), since that's the pattern I noticed. Everything under `/hub/` uses the `extension.html` layout, which doesn't check for `index.md`, just `toc: false`.

> I tried out toc: false on that same page ^^ and it didn't make any difference - the auto-generated ToC is still present.

Oops! My bad, latest commit should have fixed that

> Our commit message guidelines are here /CONTRIBUTING.md@master#commit-message-format thanks!

Will do 👍

**Comment by coopr** [comment]
OK, nice!

I see the ToC on Hub extensions, and on Docs sub-pages, but not on any of the top level pages like http://localhost:3000/ and http://localhost:3000/enterprise/ that should not have ToC - great!

I tested that `toc: false` supresses ToC on all of Kong, Kong Enterprise, and Hub pages - they all worked perfectly!

So, I _think_ we can call this done, or at least done enough. Nice-to-haves, or at least next steps, include:

- [ ] Correct heading order on all pages (so the page starts with one H1, followed by H2, H3, H4 as necessary, but never skipping from, say, H2 to H4) - this is _really_ nice to have, as the ToC on pages with incorrect headers looks pretty bad. 
- [ ] Document how ToCs work in https://github.com/Kong/docs.konghq.com/blob/master/CONTRIBUTING.md
- [ ] Do something to fix pages that have multiple identical headings (which currently breaks the ToC) - #933 would automagically fix this
- [ ] Review and consider improving the styling of ToC text (maybe H2 should be bold, H4 should be smaller, etc). A related issue was [mentioned earlier in this PR thread](https://github.com/Kong/docs.konghq.com/pull/920#issuecomment-427271005) 
> CSS changes to the heading sizes will need to be made. Because it seems like h4 tags are being used as they generate smaller font sizes than h3. I designed toc.html to enforce proper ordering of headings so, this may cause problems if headings are out of order.
- [ ] Review and consider improving the placement of ToC on Hub pages (maybe putting the ToC in the top right corner isn't ideal, and it should be at the top of the main content, like on non-Hub pages)

Anything else we should add to that list? I'll open Issues for them as necessary... Otherwise, shall we merge and deploy this? Go Hacktoberfest!! 🎃

**Comment by allejo** [comment]
> Anything else we should add to that list? I'll open Issues for them as necessary... Otherwise, shall we merge and deploy this? Go Hacktoberfest!! 🎃

Nothing that comes to mind at the moment 😄

**Comment by allejo** [comment]
Seems like that may be an issue with redcarpet once again, see https://github.com/vmg/redcarpet/issues/615.

So I originally misunderstood that this site was being compiled by GH Pages hence using my Liquid-only solution. But if you're compiling the site yourself and just hosting on GH Pages, then a plugin such as [toshimaru/jekyll-toc](https://github.com/toshimaru/jekyll-toc) instead could possibly solve a lot of these problems caused by redcarpet's generation.

Otherwise, a switch to kramdown would solve the problem with my implementation. Or ignoring headings 5 and 6 from the TOC could be a temporary workaround until it's updated to kramdown.

**Comment by coopr** [comment]
OK, thanks @allejo - I've created https://github.com/Kong/docs.konghq.com/issues/942 to track that issue.

**Comment by coopr** [comment]
I'm asking for another set of 👀 on this, as it is a rather large changeset.

I've manually reviewed and corrected all the ToCs for all Hub pages, and for Kong 0.14 and Kong Enterprise 0.33 pages - none of them are broken now. 

ToCs for pages for versions prior to 0.14 and 0.33 have not been reviewed or corrected - they'll need to be, hopefully soon.

**Comment by hutchic** [comment]
I'm still seeing some mal-formed TOC's similar to what @coopr noticed earlier

screenshot from 2018-10-16 10-42-05

quickstart on 0.2.x - 0.13.x
clustering on 0.11.x - 0.13.x
configuration on 0.10.x, 0.12.x, 0.13.x
cli on 0.10.x - 0.13.x

**Comment by coopr** [comment]
@hutchic indeed older docs need to be tuned-up - I mentioned: 

> ToCs for pages for versions prior to 0.14 and 0.33 have not been reviewed or corrected - they'll need to be, hopefully soon.

These are typically the result of mis-ordered heading levels (eg. H1 -> H3, or H2 -> H4) - I've not yet figured out a way of using find and replace to fix all of these at once, so the "repair" process is, at least for me, manual and tedious. Your contributions are welcome :)

**Comment by coopr** [comment]
Congrats @allejo - I've deployed this to production! Thanks again for your huge Hacktoberfest contribution!! If you haven't already, be sure to [claim your contributor t-shirt](https://github.com/Kong/docs.konghq.com/blob/master/CONTRIBUTING.md#contributor-t-shirt).

**Comment by coopr** [comment]
@allejo sorry, can you please remind me: Is there an option to force a ToC to appear at a particular place in the text in a given page (rather than at the default top-of-the-page location)?

**Comment by allejo** [comment]
@coopr oh no worries! There currently isn't an option to move the TOC anywhere else. On doc pages, it'll be at the top and on `hub/` it'll be on the sidebar. What did you have in mind?

**Comment by coopr** [comment]
Ah ok @allejo thanks for clarifying, much appreciated. My question came up as I was reviewing some pages that had introductory content that might have been nice to put above the ToC. Instead, I created new "Introduction" headers, and put the introductory content under those headers - problem solved!
